rfcs/0007-pass-spec.md

```diff
@@ -0,0 +1,463 @@
+If so, it should add it to the set of widgets accepting focus.
+
+Then when the widget is deleted (or disabled), it should be removed from that set.
+
+How Masonry keeps track of which widgets accept focus is outside the scope of this RFC.
+
+### Send lifecycle events directly from RenderRoot
+
+The following passes should be created in RenderRoot, which will send LifecycleEvents directly to the concerned widgets:
+
+- `UPDATE_TREE`: Updates internal flags and sends RequestPanToChild.
+- `UPDATE_FOCUS`: Sends FocusChanged event.
+- `UPDATE_DISABLED`: Sends DisabledChanged event.
+- `UPDATE_ANIM`: Sends AnimFrame event.
+- `UPDATE_SCROLLS`: Sends RequestPanToChild.
+- `UPDATE_POINTER`: Sends HotChanged and updates the cursor icon.
+
+These features should progressively be moved out of the current `WidgetPod::lifecycle` method.
+
+### Replace Lifecycle and StatusChange with StatusUpdate event
+
+The previous sections are about progressive changes that should be implemented to take things off the lifecycle pass.
+
+The lifecycle pass and the Lifecycle event should eventually be replaced with the UPDATE passes and StatusUpdate event altogether.
+
+### Rework paint pass
+
+The paint pass should now build the root scene in a single pass.
+
+`WidgetPod::paint()` will do the following:
+
+- Call the inner paint().
+- That paint "returns" both a scene fragment and an optional clip layer.
+- Append the scene fragment to the global scene.
+- If the clip layer is Some, push it.
+- For each child, call `WidgetPod::paint()` with the transform that was set in `compose()`.
+- If the clip layer is Some, pop it.
+
+Unlike the current algorithm where fragments are added to parent fragments, here they are added directly to the root.
+This is `O(N)` [instead of `O(N * Depth)`](https://xi.zulipchat.com/#narrow/stream/354396-xilem/topic/The.20mystery.20of.20the.20Xilem.20bomb).
+
+We should add `PaintCtx::set_clip_box()` and `PaintCtx::remove_clip_box()` methods so widgets can clip their children.
+In most cases this will be a rectangle, but it can be an arbitrary path, like a rectangle with rounded corners.
+
+### Get full test coverage of Context methods
+
+These behaviors and the invariants that come with them should be part of a spec.
+This spec should take the form of both doc comments, code comments, and a full testing suite.
+
+This suite should have full coverage of all Context methods and WidgetPod methods.
+Because some context methods are macro-duplicated between context types, some tests might need to be duplicated as well.
+
+
+## Drawbacks
+
+### Restricting Widget constructors
+
+Restricting Widget constructors adds another complication people learning Masonry need to be taught about.
+
+It's a complication that won't be visible to users of higher-level frameworks like Xilem.
+But it will be visible to Xilem *maintainers*; therefore that pattern should be well-documented, with various examples of code that needs to create widgets.
```

Comment:
Maybe I can't read that out of the RFC, but what is exactly meant? Is it related to the atomic creation in section `### Change how widgets are added`?

Reply:
It is. I probably need to rewrite this section once we figure out the final widget creation API.